### DISCUSSION TITLE
minor error in ex12

### DISCUSSION THREAD

**Discussion by dhcho347** [NatLabRockies/floris#657] [Q&A]
This is not important but,

https://github.com/NREL/floris/blob/main/examples/12_optimize_yaw.py#L166
```python
      yaw_opt = yaw_opt_full * (6.0 - wind_speed) / 2.0  # Linear ramp up
```

The above code should be changed like below for ramp up, I think

```python
      yaw_opt = yaw_opt_full * ( wind_speed - 4.0 ) / 2.0  # Linear ramp up
```
or
```python
      yaw_opt = yaw_opt_full * ( wind_speed - 3.0 ) / 3.0  # Linear ramp up
```

**Comment by rafmudaf** [comment]
I'm not sure of the intended result here, but just adding some more info. In the current code for wind speed of 4 m/s, that would be:

```python
yaw_opt_full * (6.0 - 4.0) / 2.0 = yaw_opt_full * (2.0 / 2.0) = yaw_opt_full
```

However, in the first proposed change, it would be `yaw_opt_full = 0.0`.
In the second, it would be:
```python
yaw_opt_full * (4.0 - 3.0) / 3.0 = yaw_opt_full * (1.0 / 3.0) = yaw_opt_full / 3
```

**Reply by dhcho347** [reply]
https://github.com/NREL/floris/blob/main/examples/12_optimize_yaw.py#L163-L170

```python
        if (wind_speed < 4.0) | (wind_speed > 14.0):
            yaw_opt = np.zeros(nturbs)  # do nothing for very low/high speeds
        elif wind_speed < 6.0:
            yaw_opt = yaw_opt_full * (6.0 - wind_speed) / 2.0  # Linear ramp up
        elif wind_speed > 12.0:
            yaw_opt = yaw_opt_full * (14.0 - wind_speed) / 2.0  # Linear ramp down
        else:
            yaw_opt = yaw_opt_full  # Apply full offsets between 6.0 and 12.0 m/s
```

□ Above code change the yaw offset on (4, 5) and (13,14) m/s
- 
as your calculation,
-   - 4m/s : (6-4)/2 = **1.0**  * yaw_opt_full
     - 5m/s : (6-5)/2 = **0.5** * yaw_opt_full .. (1.0 > 0.5 : Down) **(?)**
-   
  - 13m/s: (14-13)/2 = 0.5 * yaw_opt_full
  - 14m/s : (14-14)/2 = 0.0 * yaw_opt_full .. (0.5 > 0 : Down)
  
□ Both are down as below figure.
- 
![image](https://github.com/NREL/floris/assets/70368625/11bf983b-fe72-41f3-8694-046b539dc433)


□ I though your intention was (Up and Down) as below
- 
![image](https://github.com/NREL/floris/assets/70368625/ef217f5b-9e8b-483b-a83b-bf9f791b3266)

**Reply by rafmudaf** [reply]
Excellent results @dhcho347! Your third option looks correct to me. @bayc is the original author and will have a better idea of the intended control settings.

**Reply by Bartdoekemeijer** [reply]
Good catch! It should be the third option!

**Reply by Bartdoekemeijer** [reply]
Actually, a more elaborate answer is the following. The yaw angles should be _clipped_ rather than _scaled_ to stay within the bounds. Thus, the maximum yaw angle may be 20 deg at 12 m/s, 10 deg at 13 m/s and then 0 deg at 14 m/s. If the optimizer found an optimal yaw angle of 8 deg at 13 m/s, then this value should not be scaled or clipped: it's still within the bounds. In the current example 12, the angle is scaled (unnecessarily).

FLASC contains a code that takes care of ramp-up and ramp-down clipping. See https://github.com/NREL/flasc/blob/e425509e551c521e895c34e29db17f97ddf6f174/flasc/wake_steering/lookup_table_tools.py#L18. For more professionalized use cases, I would recommend using this function.
